File: kotlin/src/main/kotlin/com/github/fidgetting/kotlin/Jooq.kt
package com.github.fidgetting.kotlin

import com.google.protobuf.Message
import com.google.protobuf.util.JsonFormat
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.reactive.asFlow
import kotlinx.coroutines.reactive.awaitFirst
import kotlinx.coroutines.reactive.awaitFirstOrNull
import org.jooq.Condition
import org.jooq.Field
import org.jooq.JSONB
import org.jooq.Publisher

object Jooq {

  val parser = JsonFormat.parser().ignoringUnknownFields()
  val printer = JsonFormat.printer().omittingInsignificantWhitespace()

  fun <T: Message.Builder> JSONB.toProto(builder: T): T =
    builder.also { parser.merge(this.data(), it) }

  fun Message.toJSONB(): JSONB =
    JSONB.valueOf(printer.print(this))

  val <T: Any> Publisher<T>.reactive get() =
    this as org.reactivestreams.Publisher<T>

  fun <T: Any> Publisher<T>.asFlow(): Flow<T> =
    reactive.asFlow()

  suspend fun <T: Any> Publisher<T>.awaitFirst(): T =
    reactive.awaitFirst()

  suspend fun <T: Any> Publisher<T>.awaitFirstOrNull(): T? =
    reactive.awaitFirstOrNull()

  infix fun <T> Collection<T>.contains(field: Field<T>): Condition =
    field.`in`(this)
}

File: scala/src/main/scala/com/github/fidgetting/scala/Jooq.scala
package com.github.fidgetting.scala

import com.google.protobuf.Message
import com.google.protobuf.Message.Builder
import com.google.protobuf.util.JsonFormat
import org.jooq.JSONB
import org.reactivestreams.Publisher
import reactor.core.publisher.{Flux, Mono}

object Jooq {

  lazy val parser = JsonFormat.parser().ignoringUnknownFields()
  lazy val printer = JsonFormat.printer().omittingInsignificantWhitespace()

  extension (message: Message) {
    def toJSONB: JSONB = JSONB.valueOf(printer.print(message))
  }

  extension (jsonb: JSONB) {
    def toProto[T <: Message.Builder](builder: T): T = {
      parser.merge(jsonb.data(), builder)
      builder
    }
  }

  extension[T] (publisher: Publisher[T]) {
    def mono: Mono[T] = Mono.from(publisher)
    def flux: Flux[T] = Flux.from(publisher)
  }

}
